Added an option to suppress exception details

dwat001 opened the pull request:
As per comments in the Issue, this PR deals with both PostgresException.Details and the logging of command parameters.

This is related to PR https://github.com/npgsql/npgsql/pull/3008 which does the same for the 3.X branch.

YohDeadfall commented:
@dwat001 I forgot that we use `dev` instead of `master` as an active branch. I changed the target, but could you rebase your changes? Sorry for this.

dwat001 commented:
> @dwat001 I forgot that we use `dev` instead of `master` as an active branch. I changed the target, but could you rebase your changes? Sorry for this.

@YohDeadfall I have rebased onto `dev`, which branch should I be targeting to get this onto the 4.x line as well as dev? 
(Sorry for the confusion around 3.x vs 4.x)

YohDeadfall commented:
Thanks!

After merging this, the changes will appear in 5.0. To have them in 4.x versions we will backport/cherry-pick commits on our side, so no additional actions required for you.

dwat001 commented:
Thanks @YohDeadfall, could you please ping #2501 or this PR once it is cherry-picked? 
If there is back-porting to do I would be happy to do that. 

For clarity is there any further work required on this PR for it to be ready for merging?

Thank you.

YohDeadfall commented:
On backporting we add a comment to specify a cherry pick commit and the corresponding branch.

YohDeadfall commented on src/Npgsql/BackendMessages/ErrorOrNoticeMessage.cs:
```diff
@@ -58,6 +58,10 @@ internal static ErrorOrNoticeMessage Load(NpgsqlReadBuffer buf)
                     break;
                 case ErrorFieldTypeCode.Detail:
                     detail = buf.ReadNullTerminatedStringRelaxed();
+                    if(suppressDetailInPostgressError && string.IsNullOrEmpty(detail) == false)
```
It would be better to use the negation operator here rather than comparison, but it's a bit.

YohDeadfall commented:
Backported to 4.1.4 via 6fe5cde89e62912fd7763f114a921d11a7090eee.

YohDeadfall commented:
To 4.0.11 it was backported via 3f817463dd52d89feb7070a40d84fb34a6151702.

dwat001 commented:
Thank you for your time on this @YohDeadfall .

roji commented on src/Npgsql/NpgsqlCommand.cs:
```diff
@@ -1307,15 +1307,15 @@ internal void FixupRowDescription(RowDescriptionMessage rowDescription, bool isF
                 sb.Append("\t").AppendLine(s.SQL);
                 var p = s.InputParameters;
-                if (NpgsqlLogManager.IsParameterLoggingEnabled && p.Count > 0)
+                if (NpgsqlLogManager.IsParameterLoggingEnabled && p.Count > 0 && excludeParametersFromLogs == false)
```
This means that NpgsqlLogManager.IsParameterLoggingEnabled has to be enabled regardless of what's on the connection string - that seems to make the new connection string option pretty useless. I think it should be possible to simply enable parameter logging via one or the other mechanism.
